### DISCUSSION TITLE
Is it okay to name the ":isa" attribute for single inheritance?

### DISCUSSION THREAD

**Discussion by yuki-kimoto** [Ovid/Cor#72] [Q&A] [answered]
Is it okay to name the ":isa" attribute for single inheritance?

https://github.com/Ovid/Cor/blob/master/rfc/classes.md#422-inheritance

```
class Customer :isa(Person 3.14) :version(v2.1.0) {
    field $customer_id :param;

    method name :overrides () {
        my $name = $self->next::method();
        $name .= " (#$customer_id)";
        return $name;
    }
}
```

The reason I ask this question is Perl has already "isa" operator.

```
$object isa Foo
```

"isa" operator check @ISA in both single inheritance and multiple inheritance.

And if the role is implemented internally using multiple inheritance, the result of "isa" operator maybe be true.

```
class MyClass :isa(Parent) :does(Role1, Role2) {

}

my $object = MyClass->new;

if ($object isa Role1) {
  # What result? I think it maybe true.
}
```

people may ask. "Why does the isa operator return true even though the ":isa" attribute is not specified?"

**Comment by HaraldJoerg** [comment, marked as answer]
In short: It is ok to name the ":isa" attribute for single inheritance.

Both the attribute and the operator can be read like "is a" in English to give the correct understanding.  A customer is a person - makes sense.  "is a" expects singular terms on both sides.

In your example `if ($object isa Role1)` the code will _not_ return true: `Role1` is not a class, and not added to the `@ISA` of `MyClass`. With Corinna this could be caught as a syntax error because Corinna roles identify themselves at compile time.  Applied to a role implemented with `Moose::Role`, `isa Role` simply returns false.

A corresponding `does` operator in Perl does not _yet_ exist because Perl core does not have roles (yet).

**Reply by yuki-kimoto** [reply]
Thanks.

> In your example if ($object isa Role1) the code will not return true: Role1 is not a class

> Role1 is not a class, and not added to the @ISA of MyClass.

I understood now.
